In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder
df = pd.read_csv('Alphabets_data.csv')

In [2]:
df.shape

(20000, 17)

In [6]:
df.isnull().sum()

letter    0
xbox      0
ybox      0
width     0
height    0
onpix     0
xbar      0
ybar      0
x2bar     0
y2bar     0
xybar     0
x2ybar    0
xy2bar    0
xedge     0
xedgey    0
yedge     0
yedgex    0
dtype: int64

In [7]:
df['letter'].unique()

<StringArray>
['T', 'I', 'D', 'N', 'G', 'S', 'B', 'A', 'J', 'M', 'X', 'O', 'R', 'F', 'C',
 'H', 'W', 'L', 'P', 'E', 'V', 'Y', 'Q', 'U', 'K', 'Z']
Length: 26, dtype: str

In [8]:
le = LabelEncoder()
df['letter'] = le.fit_transform(df['letter'])
X = df.drop('letter', axis=1)
y = df['letter']
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

In [9]:
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score, classification_report
ann = MLPClassifier(hidden_layer_sizes=(50, 50), max_iter=500, random_state=42)
ann.fit(X_train, y_train)
y_pred = ann.predict(X_test)
print(f"Accuracy: {accuracy_score(y_test, y_pred) * 100:.2f}%")

Accuracy: 95.15%


In [10]:
from sklearn.model_selection import GridSearchCV
param_grid = {
    'hidden_layer_sizes': [(50, 50), (100,)],
    'activation': ['relu', 'tanh'],
    'learning_rate_init': [0.01, 0.001]
}
grid = GridSearchCV(MLPClassifier(max_iter=500, random_state=42), param_grid, cv=3, n_jobs=-1)

print("Tuning chal rahi hai, thoda time lagega...")
grid.fit(X_train, y_train)
print("Best Parameters:", grid.best_params_)
best_model = grid.best_estimator_
y_pred_tuned = best_model.predict(X_test)
print(f"Tuned Accuracy: {accuracy_score(y_test, y_pred_tuned) * 100:.2f}%")

Tuning chal rahi hai, thoda time lagega...
Best Parameters: {'activation': 'tanh', 'hidden_layer_sizes': (100,), 'learning_rate_init': 0.001}
Tuned Accuracy: 95.88%


C:\Users\Vishal Parmar\anaconda34\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:785: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(


In [11]:
from sklearn.metrics import classification_report

print("--- Basic Model Report ---")
print(classification_report(y_test, y_pred))

print("\n--- Tuned Model Report ---")
print(classification_report(y_test, y_pred_tuned))

--- Basic Model Report ---
              precision    recall  f1-score   support

           0       0.99      0.99      0.99       149
           1       0.93      0.95      0.94       153
           2       0.92      0.91      0.92       137
           3       0.95      0.93      0.94       156
           4       0.92      0.94      0.93       141
           5       0.94      0.94      0.94       140
           6       0.91      0.89      0.90       160
           7       0.92      0.91      0.92       144
           8       0.97      0.95      0.96       146
           9       0.94      0.97      0.95       149
          10       0.90      0.94      0.92       130
          11       0.94      0.95      0.95       155
          12       0.98      0.98      0.98       168
          13       0.96      0.93      0.94       151
          14       0.94      0.96      0.95       145
          15       0.98      0.98      0.98       173
          16       0.98      0.95      0.97       166


### Final Evaluation Report: ANN Classification (Alphabets Dataset)

#### 1. Performance Comparison
| Metric | Basic ANN Model | Tuned ANN Model |
| :--- | :--- | :--- |
| **Accuracy** | 95.15% | 95.88% |
| **Precision (Macro Avg)** | 0.95 | 0.96 |
| **Recall (Macro Avg)** | 0.95 | 0.96 |
| **F1-Score (Macro Avg)** | 0.95 | 0.96 |

**Discussion:** The tuned model achieved a higher accuracy and better balanced precision/recall across all 26 classes. The increase of ~0.73% might seem incremental, but in character recognition, this represents significantly fewer misclassifications among similar-looking alphabets.

#### 2. Tuning Methodology
We adopted a structured approach using `GridSearchCV` with 3-fold cross-validation to systematically evaluate the model's architecture. 
* **Parameters Tested:** * `hidden_layer_sizes`: [(50, 50), (100,)]
    * `activation`: ['relu', 'tanh']
    * `learning_rate_init`: [0.01, 0.001]
* **Best Configuration Found:** The grid search identified `{'activation': 'tanh', 'hidden_layer_sizes': (100,), 'learning_rate_init': 0.001}` as the optimal configuration for this dataset.

#### 3. Evaluation Analysis
* **Precision, Recall, F1-Score:** These metrics were calculated to ensure the model doesn't just perform well on majority classes but maintains consistency across all alphabet categories. The macro-average scores of 0.96 for the tuned model confirm that the model is performing uniformly well across all alphabet labels.
* **Why Tuning Matters:** Default hyperparameters often underfit complex datasets. By using `tanh` activation and a single wider hidden layer (100 neurons), the model captured non-linear spatial dependencies more effectively than the default setup.

**Conclusion:**
Hyperparameter tuning proved essential for maximizing the network's predictive performance. The optimized ANN model is robust, achieving a high final accuracy of 95.88%, confirming that the architecture is well-suited for high-dimensional character pattern recognition tasks.